pydantic/type_adapter.py

```diff
@@ -164,7 +167,12 @@ def __init__(self, type: Any, *, config: ConfigDict | None = None, _parent_depth
         try:
             core_schema = _getattr_no_parents(type, '__pydantic_core_schema__')
         except AttributeError:
-            core_schema = _get_schema(type, config_wrapper, parent_depth=_parent_depth + 1)
+            types_namespace = None
+            if hasattr(type, '__module__'):
+                types_namespace = sys.modules[type.__module__].__dict__
+            core_schema = _get_schema(
+                type, config_wrapper, parent_depth=_parent_depth + 1, types_namespace=types_namespace
+            )
```

Comment:
I think if we just make `types_namespace` an argument to the `TypeAdapter.__init__` that might be enough for users/libraries to pick up the baton. In particular, I think it would be enough that FastAPI could resolve the issue by taking responsibility for collecting the namespace when the route decorators are called.

I think it might make sense to just add that argument to the `__init__` and merge this, though it wouldn't resolve the reported issue without more tweaks. But I think this change doesn't actually _resolve_ the reported issue anyway, just addresses one instance of it.